tmail-backend/blob/secondary-blob-store/src/main/java/com/linagora/tmail/blob/secondaryblobstore/SecondaryBlobStoreDAO.java

```diff
@@ -0,0 +1,116 @@
+package com.linagora.tmail.blob.secondaryblobstore;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Collection;
+
+import org.apache.james.blob.api.BlobId;
+import org.apache.james.blob.api.BlobStoreDAO;
+import org.apache.james.blob.api.BucketName;
+import org.apache.james.blob.api.ObjectNotFoundException;
+import org.apache.james.blob.api.ObjectStoreIOException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.io.ByteSource;
+
+import jakarta.inject.Inject;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import spark.utils.IOUtils;
+
+public class SecondaryBlobStoreDAO implements BlobStoreDAO {
+    private static final Logger LOGGER = LoggerFactory.getLogger(SecondaryBlobStoreDAO.class);
+
+    private final BlobStoreDAO firstBlobStoreDAO;
+    private final BlobStoreDAO secondBlobStoreDAO;
+
+    @Inject
+    public SecondaryBlobStoreDAO(BlobStoreDAO firstBlobStoreDAO, BlobStoreDAO secondBlobStoreDAO) {
+        this.firstBlobStoreDAO = firstBlobStoreDAO;
+        this.secondBlobStoreDAO = secondBlobStoreDAO;
+    }
+
+    @Override
+    public InputStream read(BucketName bucketName, BlobId blobId) throws ObjectStoreIOException, ObjectNotFoundException {
+        try {
+            return firstBlobStoreDAO.read(bucketName, blobId);
+        } catch (Exception ex) {
+            LOGGER.error("Fail to read from the first blob store with bucket name {} and blobId {}. Use second blob store", bucketName.asString(), blobId.asString(), ex);
+            return secondBlobStoreDAO.read(bucketName, blobId);
+        }
+    }
+
+    @Override
+    public Mono<InputStream> readReactive(BucketName bucketName, BlobId blobId) {
+        return Mono.from(firstBlobStoreDAO.readReactive(bucketName, blobId))
+            .onErrorResume(ex -> {
+                LOGGER.error("Fail to read from the first blob store with bucket name {} and blobId {}. Use second blob store", bucketName.asString(), blobId.asString(), ex);
+                return Mono.from(secondBlobStoreDAO.readReactive(bucketName, blobId));
+            });
+    }
```

Comment:
Stupid question: what happens if I encounter an error while reading the inputStream?

(I'm not saying there is a good answer here, but this would definitly be a limitation of our implementation. Call sites would need to be identified and impact documented)

Reply:
updated to only catch ObjectStoreException. If firstBlobStoreDAO encounter an error while reading the inputStream, there will be no read attempt with secondBlobStoreDAO